File: cpp/15. 3-sum.cpp
//
// Created by Rutvik Patel on 11/19/23.
//

#include "include.h"
#include <algorithm>
#include <set>
#include <unordered_map>

// solution

class Solution {
public:
    std::set<std::array<int, 3>> threeSum(std::vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        std::set<std::array<int, 3>> triples;

        for (auto f_itr = nums.begin(); f_itr < nums.end(); f_itr++) {
            auto start_itr = f_itr + 1, end_itr = nums.end() - 1;

            while (start_itr < end_itr) {
                if (*start_itr + *end_itr + *f_itr > 0)
                    end_itr--;
                else if (*start_itr + *end_itr + *f_itr < 0)
                    start_itr++;
                else {
                    triples.insert({ *f_itr, *start_itr, *end_itr });
                    start_itr++;
                }
            }
        }

        return triples;
    }
};

// main

int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::vector<int>, std::set<std::array<int, 3>>>> test_cases = {
        { { -1, 0, 1, 2, -1, -4 }, { { -1, -1, 2 }, { -1, 0, 1 } } },
        { { 0, 1, 1 }, {} },
        { { 0, 0, 0 }, { { 0, 0, 0 } } },
        { { -1, 0, 1, 0 }, { { -1, 0, 1 } } },
        { { -2, 0, 1, 1, 2 }, { { -2, 0, 2 }, { -2, 1, 1 } } },
    };

    Solution solution;

    for (auto& [args, ans] : test_cases)
        assert_m(solution.threeSum(args) == ans, "Failed!");

    log("All testcases passed!");
}

File: python/15. 3-sum.py
from typing import List, Tuple


# lc sol start
class Solution:
    def threeSum(self, nums: List[int]) -> List[Tuple[int, int, int]]:
        nums.sort()
        triples = set()

        for f_idx in range(len(nums)):
            start_idx, end_idx = f_idx + 1, len(nums) - 1

            while (start_idx < end_idx):
                if nums[start_idx] + nums[end_idx] + nums[f_idx] > 0:
                    end_idx -= 1
                elif nums[start_idx] + nums[end_idx] + nums[f_idx] < 0:
                    start_idx += 1
                else:
                    triples.add((nums[f_idx], nums[start_idx], nums[end_idx]))
                    start_idx += 1

        return list(triples)
# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [([-1, 0, 1, 2, -1, -4],), ([0, 1, 1],), ([0, 0, 0],), ([-1, 0, 1, 0],), ([-2, 0, 1, 1, 2],)],
        'outputs': [[(-1, -1, 2), (-1, 0, 1)], [], [(0, 0, 0)], [(-1, 0, 1)], [(-2, 0, 2), (-2, 1, 1)]]
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        op = solution.threeSum(*args)
        assert set(op) == set(test_cases['outputs'][idx]), \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {op}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()
